File: examples/java-api-service/src/main/java/com/example/controller/ApiController.java
package com.example.controller;

import com.example.model.Order;
import com.example.model.User;
import com.example.service.OrderService;
import com.example.service.UserService;
import java.util.List;
import java.util.Optional;

/**
 * API controller handling HTTP-like request routing.
 * Demonstrates controller → service → model call chain.
 */
public class ApiController {

    private final UserService userService;
    private final OrderService orderService;

    public ApiController(UserService userService, OrderService orderService) {
        this.userService = userService;
        this.orderService = orderService;
    }

    public String handleHealthCheck() {
        return "{\"status\": \"ok\"}";
    }

    public User handleCreateUser(String name, String email) {
        return userService.createUser(name, email);
    }

    public Optional<User> handleGetUser(String id) {
        return userService.findById(id);
    }

    public Order handleCreateOrder(String userId, double amount) {
        return orderService.createOrder(userId, amount);
    }

    public List<Order> handleGetUserOrders(String userId) {
        return orderService.getOrdersByUser(userId);
    }

    public void handleConfirmOrder(String orderId) {
        orderService.confirmOrder(orderId);
    }

    public void handleCancelOrder(String orderId) {
        orderService.cancelOrder(orderId);
    }
}

File: examples/kotlin-api-service/src/main/kotlin/com/example/controller/ApiController.kt
package com.example.controller

import com.example.service.UserService
import com.example.service.OrderService

class ApiController {
    private val userService = UserService()
    private val orderService = OrderService(userService)

    fun handleGetUser(userId: String): String {
        val user = userService.findById(userId)
        return user?.displayName() ?: "User not found"
    }

    fun handleCreateOrder(userId: String, itemId: String): String {
        val order = orderService.createOrder(userId, itemId)
        return order?.let { "Order ${it.id} created" } ?: "Failed to create order"
    }

    fun handleListOrders(userId: String): List<String> {
        return orderService.findByUser(userId).map { "${it.id}: ${it.status}" }
    }
}
